framework/src/media/audio/audio_manager.c

```diff
@@ -520,74 +521,129 @@ int start_audio_stream_in(void *data, uint32_t frames)
 
 			ret = pcm_readi(pcm, cur_card->resample.readbuffer, frames_to_read);
 			medvdbg("Read frames (%d/%u) for resample to %u\n", ret, frames_to_read, frames);
-
-			int frames_total = ret;
-			int frames_used = 0;
-			int frames_copied = 0;
-			src_data_t srcData = {0, };
-
-			while (frames_total > frames_used) {
-				srcData.data_in = (void *)(cur_card->resample.readbuffer) + get_input_frames_byte_size(frames_used);
-				srcData.input_frames = frames_total - frames_used;
-				srcData.channels_num = 2;
-				srcData.origin_sample_rate = cur_card->resample.from;
-				srcData.origin_sample_width = SAMPLE_WIDTH_16BITS;
-				srcData.desired_sample_rate = cur_card->resample.to;
-				srcData.desired_sample_width = SAMPLE_WIDTH_16BITS;
-				srcData.input_frames_used = 0;
-
-				if (src_simple(cur_card->resample.handle, &srcData) != SRC_ERR_NO_ERROR) {
-					meddbg("Fail to resample to %u from %u\n", srcData.desired_sample_rate, srcData.origin_sample_rate);
-					return AUDIO_MANAGER_RESAMPLE_FAIL;
-				}
-
-				ret = get_input_frames_byte_size(srcData.output_frames_gen);
-				if (ret > 0) {
-					memcpy(data + get_input_frames_byte_size(frames_copied), srcData.data_out, ret);
-					frames_copied += srcData.output_frames_gen;
-					frames_used += srcData.input_frames_used;
-
-					medvdbg("Record resampled %d/%d\n", frames_used, frames_total);
-				} else {
-					meddbg("Failed to copy recorded data, get_input_frames_byte_size : %d\n", ret);
-					return AUDIO_MANAGER_RESAMPLE_FAIL;
-				}
-			}
-			ret = frames_copied;
 		} else {
 			ret = pcm_readi(pcm, data, frames);
 		}
 
 		medvdbg("Read %d frames\n", ret);
 		if (ret == -EPIPE) {
 			ret = pcm_prepare(pcm);
+			meddbg("PCM is reprepared\n");
 			if (ret != OK) {
 				meddbg("Fail to pcm_prepare()\n");
 				return AUDIO_MANAGER_XRUN_STATE;
 			}
 		} else if (ret == -EINVAL) {
 			return AUDIO_MANAGER_INVALID_PARAM;
 		} else {
-			return ret;
+			break;
 		}
 	} while ((ret == OK) && (retry--));
 
+	if (cur_card->resample.necessary) {
```

Comment:
You can make one function for this, too.

Reply:
Modified.